File: Shared/WorkspaceContracts.vb
Namespace Core

  Public Enum WorkspaceState
    Idle
    Capturing
    Presenting
    Dismissing
  End Enum

  Public Enum WorkspacePermissionState
    Granted
    NotDetermined
    Denied
    Restricted
  End Enum

  Public Enum WorkspaceFailureCode
    PermissionDenied = 1
    TargetDisplayUnavailable
    CaptureFailed
    CaptureCancelled
    PresentationFailed
    PlatformActualFailed
  End Enum

  Public Enum WorkspaceCommandKind
    Dismiss
    ScrollZoom
    Magnify
    Pan
    ResetScale
    ResetWorkspace
    Center
    PresetScale
    ToggleHorizontalFlip
  End Enum

  ' A frame is owned by Shared until presentFrame reports success. The surface then
  ' owns it exclusively and must release it when the presentation is dismissed.
  Public Interface IWorkspaceFrame
    Sub releaseFrame()
  End Interface

  Public Class WorkspaceFailure
    Public Property Code As WorkspaceFailureCode
    Public Property Message As String

    Public Sub New(code As WorkspaceFailureCode, message text As String)
      Me.Code = code
      Me.Message = text
    End Sub

    Public Shared Function defaultMessageForCode(code As WorkspaceFailureCode) As String
      Select Case code
        Case WorkspaceFailureCode.PermissionDenied
          Return "需要屏幕录制权限"
        Case WorkspaceFailureCode.TargetDisplayUnavailable
          Return "无法找到当前显示器"
        Case WorkspaceFailureCode.CaptureCancelled
          Return "截屏已取消"
        Case WorkspaceFailureCode.PresentationFailed
          Return "无法显示工作区"
        Case WorkspaceFailureCode.PlatformActualFailed
          Return "平台工作区实现不可用"
        Case Else
          Return "无法截取当前显示器"
      End Select
    End Function

    Public Shared Function withDefaultMessageForCode(code As WorkspaceFailureCode) As WorkspaceFailure
      Return New WorkspaceFailure(code, message: defaultMessageForCode(code))
    End Function
  End Class

  Public Class WorkspaceDisplay
    Public Property DisplayId As String
    Public Property OriginX As Double
    Public Property OriginY As Double
    Public Property Width As Double
    Public Property Height As Double
    Public Property BackingScale As Double

    Public Sub New(displayId As String, originX x As Double, originY y As Double, width displayWidth As Double, height displayHeight As Double, backingScale scale As Double)
      Me.DisplayId = displayId
      Me.OriginX = x
      Me.OriginY = y
      Me.Width = displayWidth
      Me.Height = displayHeight
      Me.BackingScale = scale
    End Sub
  End Class

  Public Class WorkspaceTransform
    Public Property Scale As Double
    Public Property OffsetX As Double
    Public Property OffsetY As Double
    Public Property IsHorizontallyFlipped As Boolean

    Public Sub New(scale As Double, offsetX x As Double, offsetY y As Double, horizontallyFlipped flipped As Boolean)
      Me.Scale = scale
      Me.OffsetX = x
      Me.OffsetY = y
      Me.IsHorizontallyFlipped = flipped
    End Sub

    Public Shared Function identityTransform() As WorkspaceTransform
      Return New WorkspaceTransform(1.0, offsetX: 0.0, offsetY: 0.0, horizontallyFlipped: False)
    End Function
  End Class

  Public Class WorkspaceRenderState
    Public Property Transform As WorkspaceTransform
    Public Property ShowsHud As Boolean

    Public Sub New(transform As WorkspaceTransform, showHud hud As Boolean)
      Me.Transform = transform
      Me.ShowsHud = hud
    End Sub
  End Class

  Public Class WorkspaceCaptureResult
    Public Property RequestId As Int64
    Public Property Frame As IWorkspaceFrame
    Public Property Display As WorkspaceDisplay
    Public Property Failure As WorkspaceFailure

    Public Sub New(requestId As Int64, frame capturedFrame As IWorkspaceFrame, onDisplay display As WorkspaceDisplay, failure failureValue As WorkspaceFailure)
      Me.RequestId = requestId
      Me.Frame = capturedFrame
      Me.Display = display
      Me.Failure = failureValue
    End Sub

    Public Property IsSuccess As Boolean
      Get
        Return assigned(Me.Frame()) AndAlso assigned(Me.Display()) AndAlso Not assigned(Me.Failure())
      End Get
    End Property

    Public Shared Function succeededWithRequestId(requestId As Int64, frame capturedFrame As IWorkspaceFrame, onDisplay display As WorkspaceDisplay) As WorkspaceCaptureResult
      Return New WorkspaceCaptureResult(requestId, frame: capturedFrame, onDisplay: display, failure: Null)
    End Function

    Public Shared Function failedWithRequestId(requestId As Int64, failure failureValue As WorkspaceFailure) As WorkspaceCaptureResult
      Return New WorkspaceCaptureResult(requestId, frame: Null, onDisplay: Null, failure: failureValue)
    End Function
  End Class

  Public Class WorkspacePresentationResult
    Public Property IsSuccess As Boolean
    Public Property Failure As WorkspaceFailure

    Public Sub New(isSuccess As Boolean, failure failureValue As WorkspaceFailure)
      Me.IsSuccess = isSuccess
      Me.Failure = failureValue
    End Sub

    Public Shared Function succeeded() As WorkspacePresentationResult
      Return New WorkspacePresentationResult(True, failure: Null)
    End Function

    Public Shared Function failedWithFailure(failure failureValue As WorkspaceFailure) As WorkspacePresentationResult
      Return New WorkspacePresentationResult(False, failure: failureValue)
    End Function
  End Class

  Public Class WorkspaceCommand
    Public Property Kind As WorkspaceCommandKind
    Public Property FirstValue As Double
    Public Property SecondValue As Double
    Public Property ThirdValue As Double

    Private Sub New(kind As WorkspaceCommandKind, firstValue first As Double, secondValue second As Double, thirdValue third As Double)
      Me.Kind = kind
      Me.FirstValue = first
      Me.SecondValue = second
      Me.ThirdValue = third
    End Sub

    Public Shared Function dismiss() As WorkspaceCommand
      Return New WorkspaceCommand(WorkspaceCommandKind.Dismiss, firstValue: 0.0, secondValue: 0.0, thirdValue: 0.0)
    End Function

    Public Shared Function scrollZoomWithDelta(delta As Double, atX x As Double, atY y As Double) As WorkspaceCommand
      Return New WorkspaceCommand(WorkspaceCommandKind.ScrollZoom, firstValue: delta, secondValue: x, thirdValue: y)
    End Function

    Public Shared Function magnifyWithAmount(magnification As Double, atX x As Double, atY y As Double) As WorkspaceCommand
      Return New WorkspaceCommand(WorkspaceCommandKind.Magnify, firstValue: magnification, secondValue: x, thirdValue: y)
    End Function

    Public Shared Function panWithDeltaX(deltaX As Double, deltaY y As Double) As WorkspaceCommand
      Return New WorkspaceCommand(WorkspaceCommandKind.Pan, firstValue: deltaX, secondValue: y, thirdValue: 0.0)
    End Function

    Public Shared Function resetScale() As WorkspaceCommand
      Return New WorkspaceCommand(WorkspaceCommandKind.ResetScale, firstValue: 0.0, secondValue: 0.0, thirdValue: 0.0)
    End Function

    Public Shared Function resetWorkspace() As WorkspaceCommand
      Return New WorkspaceCommand(WorkspaceCommandKind.ResetWorkspace, firstValue: 0.0, secondValue: 0.0, thirdValue: 0.0)
    End Function

    Public Shared Function centerInViewport(width As Double, height viewportHeight As Double) As WorkspaceCommand
      Return New WorkspaceCommand(WorkspaceCommandKind.Center, firstValue: width, secondValue: viewportHeight, thirdValue: 0.0)
    End Function

    Public Shared Function presetScaleAtAnchor(scale As Double, atX x As Double, atY y As Double) As WorkspaceCommand
      Return New WorkspaceCommand(WorkspaceCommandKind.PresetScale, firstValue: scale, secondValue: x, thirdValue: y)
    End Function

    Public Shared Function toggleHorizontalFlip() As WorkspaceCommand
      Return New WorkspaceCommand(WorkspaceCommandKind.ToggleHorizontalFlip, firstValue: 0.0, secondValue: 0.0, thirdValue: 0.0)
    End Function
  End Class

  Public Class WorkspaceStateSnapshot
    Public Property State As WorkspaceState
    Public Property StatusText As String
    Public Property RequestId As Int64
    Public Property Transform As WorkspaceTransform

    Public Sub New(state As WorkspaceState, statusText text As String, requestId id As Int64, transform transform As WorkspaceTransform)
      Me.State = state
      Me.StatusText = text
      Me.RequestId = id
      Me.Transform = transform
    End Sub
  End Class

  Public Delegate Sub WorkspaceCaptureCompletion(result As WorkspaceCaptureResult)
  Public Delegate Sub WorkspaceStateChanged(snapshot As WorkspaceStateSnapshot)
  Public Delegate Sub WorkspaceCommandRequested(command As WorkspaceCommand)
  Public Delegate Sub WorkspaceSurfaceRequested()

  Public Interface IWorkspacePlatformActual
    Function screenRecordingPermission() As WorkspacePermissionState
    Sub requestScreenRecordingPermission()
    Sub captureDisplayWithRequestId(requestId As Int64, completion completion As WorkspaceCaptureCompletion)
    Function createWorkspaceSurface() As IWorkspaceSurfaceActual
  End Interface

  Public Interface IWorkspaceSurfaceActual
    Property CommandRequested As WorkspaceCommandRequested
    Property DismissRequested As WorkspaceSurfaceRequested
    Property TargetDisplayDisconnected As WorkspaceSurfaceRequested

    Function presentFrame(frame workspaceFrame As IWorkspaceFrame, onDisplay display As WorkspaceDisplay) As WorkspacePresentationResult
    Sub renderTransform(transform As WorkspaceTransform, showHud showHud As Boolean)
    Sub dismissPresentation()
  End Interface

  Public Shared Class WorkspaceActuals
    Private Shared _registeredPlatformActual As IWorkspacePlatformActual

    Public Shared Sub registerPlatformActual(actual As IWorkspacePlatformActual)
      If Not assigned(actual) Then
        Throw New Exception("平台工作区实现不能为空")
      End If

      If assigned(_registeredPlatformActual) Then
        Throw New Exception("平台工作区实现只能在启动时注册一次")
      End If

      _registeredPlatformActual = actual
    End Sub

    Public Shared Function createSessionUsingRegisteredPlatform() As WorkspaceSession
      If Not assigned(_registeredPlatformActual) Then
        Throw New Exception("必须先注册平台工作区实现")
      End If

      Return New WorkspaceSession(_registeredPlatformActual)
    End Function
  End Class

End Namespace

File: Shared/WorkspaceContracts.pas
namespace Core;

uses
  RemObjects.Elements.RTL;

type
  WorkspaceState = public (
    Idle,
    Capturing,
    Presenting,
    Dismissing
  );

  WorkspacePermissionState = public (
    Granted,
    NotDetermined,
    Denied,
    Restricted
  );

  WorkspaceFailureCode = public (
    PermissionDenied = 1,
    TargetDisplayUnavailable,
    CaptureFailed,
    CaptureCancelled,
    PresentationFailed,
    PlatformActualFailed
  );

  WorkspaceCommandKind = public (
    Dismiss,
    ScrollZoom,
    Magnify,
    Pan,
    ResetScale,
    ResetWorkspace,
    Center,
    PresetScale,
    ToggleHorizontalFlip
  );

  // A frame is owned by Shared until presentFrame reports success. The surface then
  // owns it exclusively and must release it when the presentation is dismissed.
  IWorkspaceFrame = public interface
    method releaseFrame;
  end;

  WorkspaceFailure = public class
  private
    fCode: WorkspaceFailureCode;
    fMessage: String;
  public
    constructor(failureCode: WorkspaceFailureCode) message(text: String);
    begin
      fCode := failureCode;
      fMessage := text;
    end;
    property Code: WorkspaceFailureCode read fCode write fCode;
    property Message: String read fMessage write fMessage;

    class method defaultMessageForCode(code: WorkspaceFailureCode): String;
    begin
      case code of
        WorkspaceFailureCode.PermissionDenied:
          result := '需要屏幕录制权限';
        WorkspaceFailureCode.TargetDisplayUnavailable:
          result := '无法找到当前显示器';
        WorkspaceFailureCode.CaptureCancelled:
          result := '截屏已取消';
        WorkspaceFailureCode.PresentationFailed:
          result := '无法显示工作区';
        WorkspaceFailureCode.PlatformActualFailed:
          result := '平台工作区实现不可用';
      else
        result := '无法截取当前显示器';
      end;
    end;
    class method withDefaultMessageForCode(code: WorkspaceFailureCode): WorkspaceFailure;
    begin
      result := new WorkspaceFailure(code) message(defaultMessageForCode(code));
    end;
  end;

  WorkspaceDisplay = public class
  private
    fDisplayId: String;
    fOriginX: Double;
    fOriginY: Double;
    fWidth: Double;
    fHeight: Double;
    fBackingScale: Double;
  public
    constructor(identifier: String) originX(x: Double) originY(y: Double) width(displayWidth: Double) height(displayHeight: Double) backingScale(scale: Double);
    begin
      fDisplayId := identifier;
      fOriginX := x;
      fOriginY := y;
      fWidth := displayWidth;
      fHeight := displayHeight;
      fBackingScale := scale;
    end;
    property DisplayId: String read fDisplayId write fDisplayId;
    property OriginX: Double read fOriginX write fOriginX;
    property OriginY: Double read fOriginY write fOriginY;
    property Width: Double read fWidth write fWidth;
    property Height: Double read fHeight write fHeight;
    property BackingScale: Double read fBackingScale write fBackingScale;
  end;

  WorkspaceTransform = public class
  private
    fScale: Double;
    fOffsetX: Double;
    fOffsetY: Double;
    fIsHorizontallyFlipped: Boolean;
  public
    constructor(initialScale: Double) offsetX(x: Double) offsetY(y: Double) horizontallyFlipped(flipped: Boolean);
    begin
      fScale := initialScale;
      fOffsetX := x;
      fOffsetY := y;
      fIsHorizontallyFlipped := flipped;
    end;
    property Scale: Double read fScale write fScale;
    property OffsetX: Double read fOffsetX write fOffsetX;
    property OffsetY: Double read fOffsetY write fOffsetY;
    property IsHorizontallyFlipped: Boolean read fIsHorizontallyFlipped write fIsHorizontallyFlipped;

    class method identityTransform: WorkspaceTransform;
    begin
      result := new WorkspaceTransform(1.0) offsetX(0.0) offsetY(0.0) horizontallyFlipped(false);
    end;
  end;

  WorkspaceRenderState = public class
  private
    fTransform: WorkspaceTransform;
    fShowsHud: Boolean;
  public
    constructor(renderTransform: WorkspaceTransform) showHud(hud: Boolean);
    begin
      fTransform := renderTransform;
      fShowsHud := hud;
    end;
    property Transform: WorkspaceTransform read fTransform write fTransform;
    property ShowsHud: Boolean read fShowsHud write fShowsHud;
  end;

  WorkspaceCaptureResult = public class
  private
    fRequestId: Int64;
    fFrame: IWorkspaceFrame;
    fDisplay: WorkspaceDisplay;
    fFailure: WorkspaceFailure;
    method get_IsSuccess: Boolean;
    begin
      result := assigned(fFrame) and assigned(fDisplay) and not assigned(fFailure);
    end;
  public
    constructor(captureRequestId: Int64) frame(capturedFrame: IWorkspaceFrame) onDisplay(captureDisplay: WorkspaceDisplay) failure(failureValue: WorkspaceFailure);
    begin
      fRequestId := captureRequestId;
      fFrame := capturedFrame;
      fDisplay := captureDisplay;
      fFailure := failureValue;
    end;
    property RequestId: Int64 read fRequestId write fRequestId;
    property Frame: IWorkspaceFrame read fFrame write fFrame;
    property Display: WorkspaceDisplay read fDisplay write fDisplay;
    property Failure: WorkspaceFailure read fFailure write fFailure;
    property IsSuccess: Boolean read get_IsSuccess;

    class method succeededWithRequestId(requestId: Int64) frame(capturedFrame: IWorkspaceFrame) onDisplay(display: WorkspaceDisplay): WorkspaceCaptureResult;
    begin
      result := new WorkspaceCaptureResult(requestId) frame(capturedFrame) onDisplay(display) failure(nil);
    end;
    class method failedWithRequestId(requestId: Int64) failure(failureValue: WorkspaceFailure): WorkspaceCaptureResult;
    begin
      result := new WorkspaceCaptureResult(requestId) frame(nil) onDisplay(nil) failure(failureValue);
    end;
  end;

  WorkspacePresentationResult = public class
  private
    fIsSuccess: Boolean;
    fFailure: WorkspaceFailure;
  public
    constructor(didSucceed: Boolean) failure(failureValue: WorkspaceFailure);
    begin
      fIsSuccess := didSucceed;
      fFailure := failureValue;
    end;
    property IsSuccess: Boolean read fIsSuccess write fIsSuccess;
    property Failure: WorkspaceFailure read fFailure write fFailure;

    class method succeeded: WorkspacePresentationResult;
    begin
      result := new WorkspacePresentationResult(true) failure(nil);
    end;
    class method failedWithFailure(failureValue: WorkspaceFailure): WorkspacePresentationResult;
    begin
      result := new WorkspacePresentationResult(false) failure(failureValue);
    end;
  end;

  WorkspaceCommand = public class
  private
    fKind: WorkspaceCommandKind;
    fFirstValue: Double;
    fSecondValue: Double;
    fThirdValue: Double;

    constructor(aKind: WorkspaceCommandKind) firstValue(first: Double) secondValue(second: Double) thirdValue(third: Double);
    begin
      fKind := aKind;
      fFirstValue := first;
      fSecondValue := second;
      fThirdValue := third;
    end;
  public
    property Kind: WorkspaceCommandKind read fKind write fKind;
    property FirstValue: Double read fFirstValue write fFirstValue;
    property SecondValue: Double read fSecondValue write fSecondValue;
    property ThirdValue: Double read fThirdValue write fThirdValue;

    class method dismiss: WorkspaceCommand;
    begin
      result := new WorkspaceCommand(WorkspaceCommandKind.Dismiss) firstValue(0.0) secondValue(0.0) thirdValue(0.0);
    end;
    class method scrollZoomWithDelta(delta: Double) atX(anchorX: Double) atY(anchorY: Double): WorkspaceCommand;
    begin
      result := new WorkspaceCommand(WorkspaceCommandKind.ScrollZoom) firstValue(delta) secondValue(anchorX) thirdValue(anchorY);
    end;
    class method magnifyWithAmount(magnification: Double) atX(anchorX: Double) atY(anchorY: Double): WorkspaceCommand;
    begin
      result := new WorkspaceCommand(WorkspaceCommandKind.Magnify) firstValue(magnification) secondValue(anchorX) thirdValue(anchorY);
    end;
    class method panWithDeltaX(deltaX: Double) deltaY(y: Double): WorkspaceCommand;
    begin
      result := new WorkspaceCommand(WorkspaceCommandKind.Pan) firstValue(deltaX) secondValue(y) thirdValue(0.0);
    end;
    class method resetScaleInViewport(width: Double) height(viewportHeight: Double): WorkspaceCommand;
    begin
      result := new WorkspaceCommand(WorkspaceCommandKind.ResetScale) firstValue(0.0) secondValue(width) thirdValue(viewportHeight);
    end;
    class method resetWorkspace: WorkspaceCommand;
    begin
      result := new WorkspaceCommand(WorkspaceCommandKind.ResetWorkspace) firstValue(0.0) secondValue(0.0) thirdValue(0.0);
    end;
    class method centerInViewport(width: Double) height(viewportHeight: Double): WorkspaceCommand;
    begin
      result := new WorkspaceCommand(WorkspaceCommandKind.Center) firstValue(width) secondValue(viewportHeight) thirdValue(0.0);
    end;
    class method presetScaleAtAnchor(scale: Double) atX(anchorX: Double) atY(anchorY: Double): WorkspaceCommand;
    begin
      result := new WorkspaceCommand(WorkspaceCommandKind.PresetScale) firstValue(scale) secondValue(anchorX) thirdValue(anchorY);
    end;
    class method toggleHorizontalFlip: WorkspaceCommand;
    begin
      result := new WorkspaceCommand(WorkspaceCommandKind.ToggleHorizontalFlip) firstValue(0.0) secondValue(0.0) thirdValue(0.0);
    end;
  end;

  WorkspaceStateSnapshot = public class
  private
    fState: WorkspaceState;
    fStatusText: String;
    fRequestId: Int64;
    fTransform: WorkspaceTransform;
  public
    constructor(workspaceState: WorkspaceState) statusText(text: String) requestId(id: Int64) transform(currentTransform: WorkspaceTransform);
    begin
      fState := workspaceState;
      fStatusText := text;
      fRequestId := id;
      fTransform := currentTransform;
    end;
    property State: WorkspaceState read fState write fState;
    property StatusText: String read fStatusText write fStatusText;
    property RequestId: Int64 read fRequestId write fRequestId;
    property Transform: WorkspaceTransform read fTransform write fTransform;
  end;

  WorkspaceCaptureCompletion = public block(&result: WorkspaceCaptureResult);
  WorkspaceStateChanged = public block(snapshot: WorkspaceStateSnapshot);
  WorkspaceCommandRequested = public block(command: WorkspaceCommand);
  WorkspaceSurfaceRequested = public block;

  IWorkspacePlatformActual = public interface
    method screenRecordingPermission: WorkspacePermissionState;
    method requestScreenRecordingPermission;
    method captureDisplayWithRequestId(requestId: Int64) completion(completion: WorkspaceCaptureCompletion);
    method createWorkspaceSurface: IWorkspaceSurfaceActual;
  end;

  IWorkspaceSurfaceActual = public interface
    property CommandRequested: WorkspaceCommandRequested read write;
    property DismissRequested: WorkspaceSurfaceRequested read write;
    property TargetDisplayDisconnected: WorkspaceSurfaceRequested read write;

    method presentFrame(workspaceFrame: IWorkspaceFrame) onDisplay(display: WorkspaceDisplay): WorkspacePresentationResult;
    method renderTransform(transform: WorkspaceTransform) showHud(showHud: Boolean);
    method dismissPresentation;
  end;

  WorkspaceActuals = public static class
  private
    class field fRegisteredPlatformActual: IWorkspacePlatformActual;
  public
    class method registerPlatformActual(actual: IWorkspacePlatformActual);
    begin
      if not assigned(actual) then
        raise new Exception('平台工作区实现不能为空');

      if assigned(fRegisteredPlatformActual) then
        raise new Exception('平台工作区实现只能在启动时注册一次');

      fRegisteredPlatformActual := actual;
    end;
    class method createSessionUsingRegisteredPlatform: WorkspaceSession;
    begin
      if not assigned(fRegisteredPlatformActual) then
        raise new Exception('必须先注册平台工作区实现');

      result := new WorkspaceSession(fRegisteredPlatformActual);
    end;
  end;

end.
